Memory optimization for MOOSE PRNG

permcody opened the pull request:
Currently we're storing all generator states per dof
but we simply don't need to do that. This PR stores
less state reducing memory usage.

~~Additionally, I'm re-enabling the random test. I believe this may be working again due to recent threading fixes.~~

refs #6590, #2088

permcody commented:
![figure_1](https://cloud.githubusercontent.com/assets/3036960/13794292/cfc988be-eac0-11e5-85dd-a68bab282864.png)

jwpeterson commented:
If you're re-enabling an otherwise-unrelated test, I'd like to see that in at *least* a separate commit, if not a separate PR entirely.

permcody committed:
Memory optimization for MOOSE PRNG

Currently we're storing all generator states per dof
but we simply don't need to do that. This PR stores
less state reducing memory usage.

closes #6590

jwpeterson commented:
:+1: thanks for making the memory_logger plots.

friedmud commented on framework/src/utils/RandomData.C:
```diff
@@ -96,32 +96,29 @@ RandomData::updateGenerators()
-  {
-    MeshBase::const_node_iterator it = _rd_mesh.getMesh().active_nodes_begin();
-    MeshBase::const_node_iterator end_it = _rd_mesh.getMesh().active_nodes_end();
+    updateGeneratorHelper(_rd_mesh.getMesh().active_nodes_begin(), _rd_mesh.getMesh().active_nodes_end());
```
Why not just loop over local nodes and elements?

permcody replied:
Because then you don't get consistent random numbers for different numbers
of processors. You need to generate the same sequence regardless of the
decomposition.
On Wed, Mar 16, 2016 at 9:50 AM Derek Gaston <notifications@github.com>
wrote:

> In framework/src/utils/RandomData.C
> <https://github.com/idaholab/moose/pull/6591#discussion_r56359020>:
>
> > @@ -96,32 +96,29 @@ RandomData::updateGenerators()
> >    }
> >
> >    if (_is_nodal)
> > -  {
> > -    MeshBase::const_node_iterator it = _rd_mesh.getMesh().active_nodes_begin();
> > -    MeshBase::const_node_iterator end_it = _rd_mesh.getMesh().active_nodes_end();
> > +    updateGeneratorHelper(_rd_mesh.getMesh().active_nodes_begin(), _rd_mesh.getMesh().active_nodes_end());
>
> Why not just loop over local nodes and elements?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6591/files/a20165253410cfe9bfc05d3c3b04e1da92707c03#r56359020>
>

friedmud replied:
Ummm... then there is no way this can work with ParallelMesh... so the whole thing needs to be redesigned.

Having a serial loop over all elements / nodes is bad news anyway.  It should be threaded (which would also preclude this design).

friedmud replied:
Wow - I just really looked at the code for the first time... the current implementation is just not right.  Relying on iterating over the whole mesh in the same way on every processor just isn't right.

It shouldn't hold up this PR (this is definitely an improvement).  But this needs to get redesigned.

permcody replied:
There are already notes in there about this being deficient for parallel
mesh. That has been known since the start. I don't think it really matters
though, it just means you'd get different numbers in that case. I'm open to
ideas but keep in mind where we came from and why we have this design.
Without the two-level generator you can end up with patterns in your random
numbers.

There is certainly an issue with scalability in time here but not space
anymore.
On Wed, Mar 16, 2016 at 10:01 AM Derek Gaston <notifications@github.com>
wrote:

> In framework/src/utils/RandomData.C
> <https://github.com/idaholab/moose/pull/6591#discussion_r56361061>:
>
> > @@ -96,32 +96,29 @@ RandomData::updateGenerators()
> >    }
> >
> >    if (_is_nodal)
> > -  {
> > -    MeshBase::const_node_iterator it = _rd_mesh.getMesh().active_nodes_begin();
> > -    MeshBase::const_node_iterator end_it = _rd_mesh.getMesh().active_nodes_end();
> > +    updateGeneratorHelper(_rd_mesh.getMesh().active_nodes_begin(), _rd_mesh.getMesh().active_nodes_end());
>
> Wow - I just really looked at the code for the first time... the current
> implementation is just not right. Relying on iterating over the whole mesh
> in the same way on every processor just isn't right.
>
> It shouldn't hold up this PR (this is definitely an improvement). But this
> needs to get redesigned.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6591/files/a20165253410cfe9bfc05d3c3b04e1da92707c03#r56361061>
>
